src/enclosure.c

```diff
@@ -171,3 +184,151 @@ int enclosure_open(const struct enclosure_device *enclosure)
+{
+	struct enclosure_device *enclosure_device;
+	list_for_each(sysfs_get_enclosure_devices(), enclosure_device) {
+		if (strncmp(enclosure_device->dev_path, enclosure_id, PATH_MAX) == 0) {
+			return enclosure_device;
+		}
+	}
+	return NULL;
+}
+
+static struct ses_slot *find_enclosure_slot_by_index(struct enclosure_device *encl, int index)
+{
+	for (int i = 0; i < encl->slots_count; i++) {
+		if (encl->slots[i].index == index) {
+			return &encl->slots[i];
+		}
+	}
+	return NULL;
+}
+
+static status_t _enclosure_get_slot(struct enclosure_device *encl, int index,
+									const char *device, struct slot_response *slot_res)
+{
+	struct block_device *block_device = NULL;
+	struct ses_slot *s_slot = find_enclosure_slot_by_index(encl, index);
+	if (!s_slot) {
+		log_error("SCSI: Unable to locate slot in enclosure %d\n", index);
+		return STATUS_NULL_POINTER;
+	}
+
+	slot_res->state = s_slot->ibpi_status;
+	snprintf(slot_res->slot, PATH_MAX, "%s/%d", encl->dev_path, index);
+
+	if (!device) {
+		block_device = locate_block_by_sas_addr(s_slot->sas_addr);
+		if (block_device) {
+			device = basename(block_device->sysfs_path);
+		}
+	}
+
+	// Not having a block device for a slot is OK
+	if (device) {
+		snprintf(slot_res->device, PATH_MAX, "/dev/%s", device);
+	}
+
+	return STATUS_SUCCESS;
+}
+
+static status_t parse_slot_id(char *slot_num, char *enclosure_id, int *index)
+{
+	char tmp_enclosure_id[PATH_MAX];
+	const char *index_str;
+
+	if (slot_num == NULL || slot_num[0] == '\0') {
+		log_error("SCSI: Invalid slot identifier =%s\n", slot_num);
+		return STATUS_NULL_POINTER;
+	}
+
+	snprintf(tmp_enclosure_id, PATH_MAX, "%s", slot_num);
+	snprintf(enclosure_id, PATH_MAX, "%s", dirname(tmp_enclosure_id));
+	index_str = basename(slot_num);
+
+	if (str_toi(index, index_str, NULL, 10)) {
+		log_error("SCSI: Invalid slot identifier index %s\n", index_str);
+		return STATUS_INVALID_PATH;
+	}
+	return STATUS_SUCCESS;
+}
+
+static status_t enclosure_get_slot_by_slot_num(char *slot_num, struct slot_response *slot_res)
+{
+	char enclosure_id[PATH_MAX];
+	int index = -1;
+	struct enclosure_device *encl = NULL;
+
+	status_t parse = parse_slot_id(slot_num, enclosure_id, &index);
+	if (STATUS_SUCCESS != parse)
+		return parse;
+
+	encl = find_enclosure(enclosure_id);
+	if (!encl) {
+		log_error("SCSI: Invalid enclosure ='%s'\n", enclosure_id);
+		return STATUS_NULL_POINTER;
+	}
+	return _enclosure_get_slot(encl, index, NULL, slot_res);
+}
+
+static status_t enclosure_get_slot_by_device(char *device, struct slot_response *slot_res)
+{
+	struct block_device *block_device = get_block_device_from_sysfs_path(basename(device));
+
+	if (!block_device) {
+		log_error("SCSI: Device node not found %s\n", device);
+		return STATUS_INVALID_PATH;
+	}
+
+	if (!block_device->enclosure) {
+		log_error("SCSI: Not a SCSI ses device %n\n", device);
+		return STATUS_INVALID_PATH;
+	}
+	return _enclosure_get_slot(block_device->enclosure, block_device->encl_index, device, slot_res);
```

Comment:
I prefer to use block device determined by application instead of using value passed by user. Could you please extract the `device` from `block_device`?

This is the programming conception to make user input meaningless as possible. We are using `device` to get `block _device` and after that `device` is no longer necessary.

Reply:
Done